gmeta/src/lib.rs

```diff
@@ -1,66 +1,189 @@
+//! The metadata informs the user about the contract's interface and allows them
+//! to interact with it using custom types.
+//!
+//! To generate a metadata output file `contract_name.meta.txt` for a contract,
+//! you need:
+//!
+//! - Add `gmeta` crate to your `Cargo.toml` file.
+//! - Define an empty struct that will identify the contract metadata.
+//! - Implement the [`Metadata`] trait for this struct by defining the
+//!   associated types of the trait.
+//! - Call [`gear_wasm_builder::build_with_metadata`](https://docs.gear.rs/gear_wasm_builder/fn.build_with_metadata.html)
+//!   function instead of [`gear_wasm_builder::build`](https://docs.gear.rs/gear_wasm_builder/fn.build.html)
+//!   function in `build.rs` file.
+//!
+//! # Example
+//!
+//! In this example we will create a simple ping-pong contract.
+//!
+//! We will define message types for `handle()` and `state()` functions.
+//!
+//! ```
+//! #[no_std]
+//! use gmeta::{InOut, Metadata};
+//! use gstd::{msg, prelude::*, ActorId};
+//!
+//! // Counter that will be incremented on each `Ping` message.
+//! static mut COUNTER: i32 = 0;
+//!
+//! // Message type for `handle()` function.
+//! #[derive(Encode, Decode, TypeInfo)]
+//! pub enum PingPong {
+//!     Ping,
+//!     Pong,
+//! }
+//!
+//! // Metadata struct.
+//! pub struct ContractMetadata;
+//!
+//! impl Metadata for ContractMetadata {
+//!     // The unit tuple is used as neither incoming nor outgoing messages are
+//!     // expected in the `init()` function.
+//!     type Init = ();
+//!     // We use the same `PingPong` type for both incoming and outgoing
+//!     // messages.
+//!     type Handle = InOut<PingPong, PingPong>;
+//!     // The unit tuple is used as we don't use asynchronous interaction in this
+//!     // contract.
+//!     type Others = ();
+//!     // The unit tuple is used as we don't process any replies in this contract.
+//!     type Reply = ();
+//!     // The unit tuple is used as we don't process any signals in this contract.
+//!     type Signal = ();
+//!     // We return a counter value (`i32`) in the `state()` function in this contract.
+//!     type State = i32;
+//! }
+//!
+//! #[no_mangle]
+//! extern "C" fn handle() {
+//!     // Load incoming message of `PingPong` type.
+//!     let payload: PingPong = msg::load().expect("Unable to load");
+//!
+//!     if let PingPong::Ping = payload {
+//!         unsafe { COUNTER += 1 };
+//!         // Send a reply message of `PingPong` type back to the sender.
+//!         msg::reply(PingPong::Pong, 0).expect("Unable to reply");
+//!     }
+//! }
+//!
+//! #[no_mangle]
+//! extern "C" fn state() {
+//!     msg::reply(unsafe { COUNTER }, 0).expect("Unable to reply");
+//! }
+//! ```
+//!
+//! `build.rs` file:
+//!
+//! ```no_run
+//! # const IGNORE: &'static str = stringify! {
+//! use ping::ContractMetadata;
+//! # };
+//! # pub struct ContractMetadata;
+//! # impl gmeta::Metadata for ContractMetadata {
```

Comment:
what's it?

Reply:
btw note that if IO mod exists metadata exactly may be formed in build.rs file

Reply:
Note that `#` in examples code is for hidden code. It is needed only for compiling examples (as they are doc texts too).

See for updates: https://github.com/gear-tech/gear/pull/2761/commits/6377f4b